core/src/main/java/org/apache/accumulo/core/clientImpl/TabletServerBatchWriter.java

```diff
@@ -639,6 +590,39 @@ public MutationWriter(int numSendThreads) {
       binningThreadPool =
           ThreadPools.createFixedThreadPool(1, "BinMutations", new SynchronousQueue<>(), false);
       binningThreadPool.setRejectedExecutionHandler(new ThreadPoolExecutor.CallerRunsPolicy());
+      binningThreadPool.execute(Trace.wrap(new Runnable() {
+        @Override
+        public void run() {
+          while (!closed) {
+            int num = incomingQueue.numMutations();
+            if (num > 0) {
+              try {
+                log.trace("{} - binning {} mutations", Thread.currentThread().getName(), num);
+                // Remove the mutations from the incoming queue
+                final MutationSet copy = new MutationSet();
+                incomingQueue.forEach((table, mutations) -> {
+                  mutations.forEach(mutation -> {
+                    copy.add(table, mutation);
+                    mutations.remove(mutation);
+                  });
+                });
+                addMutations(copy);
+              } catch (Exception e) {
+                updateUnknownErrors("Error processing mutation set", e);
+              }
+            } else {
+              try {
+                Thread.sleep(100);
```

Comment:
If incomingQueue were of type `BlockingQueue<TableIDAndMutation>` the call to sleep could be avoided using `take()` followed by `drainTo()` or `poll()` methods on  BlockingQueue instead.